ArcGISRuntimeSDKQt_CppSamples/UtilityNetwork/ValidateUtilityNetworkTopology/ValidateUtilityNetworkTopology.cpp

```diff
@@ -0,0 +1,671 @@
+
+void ValidateUtilityNetworkTopology::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<ValidateUtilityNetworkTopology>("Esri.Samples", 1, 0, "ValidateUtilityNetworkTopologySample");
+}
+
+MapQuickView* ValidateUtilityNetworkTopology::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void ValidateUtilityNetworkTopology::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_busy = true;
+  emit isBusy();
+  m_message = QString("Loading a webmap...");
+  emit messageChanged();
+
+  Envelope envelope(
+      -9815489.0660101417,
+      5128463.4221229386,
+      -9814625.2768726498,
+      5128968.4911854975,
+      SpatialReference::webMercator()
+      );
+
+  Viewpoint initialViewpoint(envelope);
+
+  m_map->setInitialViewpoint(initialViewpoint);
+
+  // Load in persistent session mode (workaround for server caching issue)
+  // https://support.esri.com/en-us/bug/asynchronous-validate-request-for-utility-network-servi-bug-000160443
+  m_map->loadSettings()->setFeatureServiceSessionType(FeatureServiceSessionType(FeatureServiceSessionType::Persistent));
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  m_overlay = new GraphicsOverlay(this);
+  m_mapView->graphicsOverlays()->append(m_overlay);
+
+  emit mapViewChanged();
+
+  connectSignals();
+}
+
+void ValidateUtilityNetworkTopology::connectSignals()
+{
+  connect(m_mapView, &MapQuickView::mouseClicked, this, [this](QMouseEvent& mouseEvent)
+  {
+    // first clear the selection
+    m_deviceFeatureLayer->clearSelection();
+    m_lineFeatureLayer->clearSelection();
+
+    m_busy = true;
+    emit isBusy();
+
+    m_message = QString("Identifying feature to edit...");
+    emit messageChanged();
+
+    if (m_map->loadStatus() != LoadStatus::Loaded)
+      return;
+
+    m_clickPoint = m_mapView->screenToLocation(mouseEvent.position().x(), mouseEvent.position().y());
+
+    // Perform an identify to determine if a user tapped on a feature.
+    m_mapView->identifyLayersAsync(mouseEvent.position(), 5.0, false).then(this, [this](const QList<IdentifyLayerResult*>& results)
+    { if(!results.isEmpty())
+      {
+        for(IdentifyLayerResult* result : results)
+        {
+          if(result && (result->layerContent()->name() == m_deviceTableName || result->layerContent()->name() == m_lineTableName))
+          {
+            m_feature = static_cast<ArcGISFeature*>(qAsConst(result)->geoElements()[0]);
+            break;
+          }
+        }
+      }
+      else
+      {
+        m_message = QString("No feature identified. Tap on a feature to edit.");
+        emit messageChanged();
+
+        m_busy = false;
+        emit isBusy();
+
+        return;
+      }
+      m_updateFieldName = m_feature->featureTable()->tableName() == m_deviceTableName ? m_deviceStatusField : m_nominalVoltageField;
```

Comment:
All this logic needs to be moved to another function. This is too much to happen within `ValidateUtilityNetworkTopology::connectSignals`. Consider an `onIdentifyLayersAsyncCompleted` function.

Reply:
done